cpp/src/copying/shift.cu

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (c) 2019-2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <algorithm>
+#include <iterator>
+#include <memory>
+#include <thrust/copy.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/transform.h>
+#include <cudf/types.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/copying.hpp>
+#include <cudf/detail/valid_if.cuh>
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/traits.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+namespace cudf {
+namespace experimental {
+namespace {
+
+inline bool __device__ out_of_bounds(size_type size, size_type idx) {
+    return idx < 0 || idx >= size;
+}
+
+template<typename T>
+struct value_functor {
+    column_device_view const input;
+    size_type size;
+    size_type offset;
+    T const* fill;
+
+    T __device__ operator()(size_type idx) {
+        auto src_idx = idx - offset;
+        return out_of_bounds(size, src_idx)
+            ? *fill
+            : input.element<T>(src_idx);
+    }
+};
+
+struct validity_functor {
```

Comment:
Can these be lambdas instead of function objects to make it easier to read at the location where they are used?

Reply:
They can be, but if I recall correctly, it requires a large capture clause, which reduced readability, so it was 🤷🏻‍♂️. I’ll take another look.